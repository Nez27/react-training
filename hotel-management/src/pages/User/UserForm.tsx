// Hooks
import { FormProvider, useForm } from 'react-hook-form';
import { useCreateUser } from '@hook/users/useCreateUser.ts';
import { useUpdateUser } from '@hook/users/useUpdateUser.ts';

// Styled
import Input from '@commonStyle/Input.ts';

// Components
import Form from '@component/Form/index.tsx';

// Helpers
import { isValidRegex } from '@helper/validators.ts';

import {
  INVALID_FIELD,
  INVALID_PHONE,
  REQUIRED_FIELD_ERROR,
} from '@constant/formValidateMessage.ts';
import { REGEX } from '@constant/commons.ts';

// Styled
import { FormBtn } from './styled.ts';

// Types
import { IUser } from '@type/users.ts';

interface IUserFormProp {
  onCloseModal?: () => void;
  user?: IUser;
}

const UserForm = ({ onCloseModal, user }: IUserFormProp) => {
  const { isCreating, createUser } = useCreateUser();
  const { isUpdating, updateUser } = useUpdateUser();
  const isLoading = isCreating || isUpdating;
  const {id: editId, ...editValues} = {...user};
  
  const formMethods = useForm<IUser>({
    defaultValues: editId 
      ? editValues 
      : {}
  });
  const {
    register,
    handleSubmit,
    reset,
    formState: { errors, isDirty, isValid },
    trigger,
  } = formMethods;

  // Submit form
  const onSubmit = 
    (newUser: IUser) => {
      if (!user) {
        // Add request
        createUser(newUser, {
          onSuccess: () => {
            reset();
            onCloseModal?.();
          },
        });
      } else {
        // Edit request
        newUser.id = editId!;
        updateUser(newUser, {
          onSuccess: () => {
            reset();
            onCloseModal?.();
          }
        });
      }
    };

  return (
    <FormProvider {...formMethods}>
      <Form onSubmit={handleSubmit(onSubmit)}>
        <Form.Row label="Full Name" error={errors?.name?.message}>
          <Input
            type="text"
            id="name"
            {...register('name', {
              required: REQUIRED_FIELD_ERROR,
              validate: {
                checkValidName: (value) =>
                  isValidRegex(new RegExp(REGEX.NAME), value) || INVALID_FIELD,
              },
              onChange: () => trigger('name'),
            })}
          />
        </Form.Row>

        <Form.Row label="Phone" error={errors?.phone?.message}>
          <Input
            type="text"
            id="phone"
            {...register('phone', {
              required: REQUIRED_FIELD_ERROR,
              validate: {
                checkPhoneNum: (v) =>
                  isValidRegex(new RegExp(REGEX.PHONE), v) || INVALID_PHONE,
              },
              onChange: () => trigger('phone'),
            })}
          />
        </Form.Row>

        <Form.Action>
          <FormBtn
            type="submit"
            name="submit"
            disabled={!isDirty || !isValid || isLoading}
          >
            {
              !user 
                ? 'Add' 
                : 'Save'
            }
          </FormBtn>
          <FormBtn type="button" variations="secondary" onClick={onCloseModal}>
            Close
          </FormBtn>
        </Form.Action>
      </Form>
    </FormProvider>
  );
};

export default UserForm;
