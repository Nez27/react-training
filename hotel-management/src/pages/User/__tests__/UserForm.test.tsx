import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { BrowserRouter } from 'react-router-dom';

jest.mock('@hook/users/useCreateUser.ts', () => ({
  ...jest.requireActual('@hook/users/useCreateUser.ts'),
  useCreateUser: () => ({
    isCreating: true,
    createUser: jest.fn(),
  }),
}));

// Components
import UserForm from '../UserForm';
import { RenderResult, fireEvent, render } from '@testing-library/react';
import { act } from 'react-test-renderer';

describe('UserForm', () => {
  const queryClient = new QueryClient();

  let wrapper: RenderResult | null = null;

  beforeEach(() => {
    wrapper = render(
      <QueryClientProvider client={queryClient}>
        <BrowserRouter>
          <UserForm />
        </BrowserRouter>
      </QueryClientProvider>
    );
  });

  test('Should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('Should submit correctly', async () => {
    const nameField = wrapper!.container.querySelector('#name');
    const phoneField = wrapper!.container.querySelector('#phone');
    const submitBtn = wrapper!.getByText('Add');

    await act(async () => {
      fireEvent.input(nameField!, { target: { value: 'Phan Huu Loi' } });
      fireEvent.input(phoneField!, { target: { value: '0324432321' } });
      fireEvent.click(submitBtn);
    });

    wrapper?.debug();
  });
});
