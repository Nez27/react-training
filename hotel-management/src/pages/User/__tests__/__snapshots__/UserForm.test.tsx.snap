// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserForm Should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="sc-dlniIP bCPNqw"
      >
        <div
          class="sc-eCAqax dgDilB"
          direction="horizontal"
        >
          <label
            class="sc-jSFipO cMYxin"
          >
            Full Name
          </label>
          <div>
            <input
              class="sc-bdnyFh gXSWSo"
              id="name"
              name="name"
              type="text"
            />
          </div>
        </div>
        <div
          class="sc-eCAqax dgDilB"
          direction="horizontal"
        >
          <label
            class="sc-jSFipO cMYxin"
          >
            Phone
          </label>
          <div>
            <input
              class="sc-bdnyFh gXSWSo"
              id="phone"
              name="phone"
              type="text"
            />
          </div>
        </div>
        <div
          class="sc-hKFymg dteuGK"
        >
          <button
            class="sc-gtsqUy sc-iCoHzw iXmfC cwhbgM"
            disabled=""
            name="submit"
            type="submit"
            variations="primary"
          >
            Add
          </button>
          <button
            class="sc-gtsqUy sc-iCoHzw hesuRJ cwhbgM"
            type="button"
            variations="secondary"
          >
            Close
          </button>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="sc-dlniIP bCPNqw"
    >
      <div
        class="sc-eCAqax dgDilB"
        direction="horizontal"
      >
        <label
          class="sc-jSFipO cMYxin"
        >
          Full Name
        </label>
        <div>
          <input
            class="sc-bdnyFh gXSWSo"
            id="name"
            name="name"
            type="text"
          />
        </div>
      </div>
      <div
        class="sc-eCAqax dgDilB"
        direction="horizontal"
      >
        <label
          class="sc-jSFipO cMYxin"
        >
          Phone
        </label>
        <div>
          <input
            class="sc-bdnyFh gXSWSo"
            id="phone"
            name="phone"
            type="text"
          />
        </div>
      </div>
      <div
        class="sc-hKFymg dteuGK"
      >
        <button
          class="sc-gtsqUy sc-iCoHzw iXmfC cwhbgM"
          disabled=""
          name="submit"
          type="submit"
          variations="primary"
        >
          Add
        </button>
        <button
          class="sc-gtsqUy sc-iCoHzw hesuRJ cwhbgM"
          type="button"
          variations="secondary"
        >
          Close
        </button>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
